### ISSUE TITLE
 Problem with orders with amounts without decimals

### ISSUE THREAD

**Issue by mastudillot** [PrestaShop/PrestaShop#26553]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

When Prestashop is configured so that it does not use decimals, when validating the order with the 'validateOrder' method, the order is marked with payment error (PS_OS_ERROR)

on line 323 the error is thrown
```php
if ($order_status->logable && number_format($cart_total_paid, Context::getContext()->getComputingPrecision()) != number_format($amount_paid, _PS_PRICE_COMPUTE_PRECISION_)) {
    $id_order_state = Configuration::get('PS_OS_ERROR');
}
```

In the if, the amount is validated with the constant _PS_PRICE_COMPUTE_PRECISION_ and the configuration of the currency is not checked. In this case the variable **_$cart_total_paid_** always has a difference with the amount paid (**_number_format($amount_paid, _PS_PRICE_COMPUTE_PRECISION_)_**).

![image](https://user-images.githubusercontent.com/36648048/140820904-b65822bc-b83f-408d-860b-94b7ed222edf.png)

### Expected behavior

The order is marked with the correct status

### Steps to reproduce

1. Install clean Prestashop versión >= 1.7.8.0.
2. Create currency without decimal and set as default.
3. Install the `Payment by check` and enable it or use a payment method that set the status of payment in `payment accepted (ps_order_state id 2)` or `Processing in progress (ps_order_state id 3)`.
3. Create an publish product with price lower than $1000.
4. Create and publish product with price higher than $1000.
5. Buy a product with the `Payment by check` payment method or the one configured with a price less than $ 1000. It should work fine.
6. Purchase product with the `Payment by check` payment method or the one configured with price higher than $1000. This will fail and set order to Payment Error state.

### PrestaShop version(s) where the bug happened

1.7.8.0

### PHP version(s) where the bug happened

7.3

**Comment by florine2623** [comment]
Hello @mastudillot ,

I followed your steps, but I can't reproduce the issue.
My own currency has been created and is set as default :

<img width="823" alt="Screenshot 2021-11-09 at 10 58 10" src="https://user-images.githubusercontent.com/16019289/140902990-867f203e-1939-4f9f-b0fb-1ae6a92af420.png">
<img width="415" alt="Screenshot 2021-11-09 at 10 58 22" src="https://user-images.githubusercontent.com/16019289/140903014-b5365832-7c5b-4cea-a261-6c68ae64dea3.png">


I placed an order :
<img width="1128" alt="Screenshot 2021-11-09 at 10 55 40" src="https://user-images.githubusercontent.com/16019289/140902847-7be744a7-a118-4c34-a4f1-5cd696ad477f.png">

I changed the status to my order to "Payment accepted" :
<img width="1177" alt="Screenshot 2021-11-09 at 10 56 09" src="https://user-images.githubusercontent.com/16019289/140902862-feae696c-b16e-4c9f-82b0-f2f6e16b2009.png">

Is the issue only coming when you're using Webpay Plus ?

Please check and feedback.
Thanks!

**Comment by mastudillot** [comment]
With the plugin webpay plus or mercado pago. 

Those plugins use the "validateOrder" method of the "PaymentModuleCore" class.

Forget to add in the steps to reproduce the problem, that the order amount must be greater than 1000.

![image](https://user-images.githubusercontent.com/36648048/140932542-86ca4cbb-8baf-4ac5-8d09-374ee6702817.png)

**Comment by florine2623** [comment]
Thanks for the details.

I tested again with order amount superior to 1 000 and 10 000. I still can't reproduce your issue.
The Order status is correct :
<img width="1167" alt="Screenshot 2021-11-09 at 17 13 08" src="https://user-images.githubusercontent.com/16019289/140961685-11f66f7f-6d6b-4b90-88e7-79447f77be05.png">

You should contact the develop of these modules if the issue only comes with those payment modules.
As you can see I can't reproduce on my end using a native payment module from PrestaShop.

So I will close your issue.
Thanks!

**Comment by mastudillot** [comment]
In the amounts of your currency the separator is "," in my tests the separator is "."

I work in the development of the webpay plugin and this problem is causing problems for prestashop users.

Can you review the last thing I comment on please. 🥺

**Comment by mastudillot** [comment]
Payment by check uses the _**order_state**_ with **_id 1_**, our plugin uses **_id 3_** when placing the order.

If in the order_state table I change the **logable column to 1**, the error that I mentioned is reproduced and it gets into the if that I mention in the report.
![image](https://user-images.githubusercontent.com/36648048/140995449-a4b12dbe-19ef-45ce-a36a-cdac87d43d39.png)

**Comment by mpaolino** [comment]
This exact same issue happens to me. 
PrestaShop 1.7.8.0 and 1.7.8.1 with [MercadoPago plugin v4.8.0](https://addons.prestashop.com/es/pago-tarjeta-carteras-digitales/23962-mercado-pago.html). 

Configured the only active currency with 0 decimals, "," as decimal separator.

As @mastudillot  already explained, the culprit is in [validateOrder of PaymentModule.php](https://github.com/PrestaShop/PrestaShop/blob/586dbdbb3c60899f9f07b1eb41864b9e4333175a/classes/PaymentModule.php#L319).

Under this situation:

* getContext()->getComputingPrecision returns 0 (changing dynamically as per our configuration)
* _PS_PRICE_COMPUTE_PRECISION_ is set to 2 in [config/config.inc.php](https://github.com/PrestaShop/PrestaShop/blob/586dbdbb3c60899f9f07b1eb41864b9e4333175a/config/config.inc.php#L141)

If you issue a $1000 order these will return the strings "1000" and "1000,00" respectively, making the conditional true.

If you take a closer look at the config file you would spot the deprecation:

![Screenshot from 2021-12-01 10-25-56](https://user-images.githubusercontent.com/757973/144242466-46b489fb-1530-4924-af43-a16c7f1ef1ed.png)

I wonder why something deprecated is used by `validateOrder` function in `classes/PaymentModule.php`. Any module that uses such functionality, as the MercadoPago official partner plugin, is going to fail under this configuration.

Also Prestashop compares two values by building their string representations with two different precision configurations, seems this was wrong from the start, even after the deprecation. 

Even more important this situation doesn't generate any error log nor any deprecation warning, **in a payment core class none the less.**

@mastudillot I temporarily fixed it by matching the value of _PS_PRICE_COMPUTE_PRECISION_ to the decimals of my currency, this only fixes for my particular case and has other issues if you use it with more than one active currencies with different configs.

**Comment by mastudillot** [comment]
@mpaolino To solve the problem in the Transbank plugin. I generated new payment statuses with the logable field at 0. With that field at 0 it does not go through the check.

**Comment by mpaolino** [comment]
@florine2623 can you please reopen this issue? This is clearly a bug.

**Comment by mpaolino** [comment]
> @mpaolino To solve the problem in the Transbank plugin. I generated new payment statuses with the logable field at 0. With that field at 0 it does not go through the check.

Seems this should be really fixed by changing the _PS_PRICE_COMPUTE_PRECISION_ by using `Context::getContext()->getComputingPrecision()` again, its the only place where that former constant is being used instead of the later.

**Comment by mpaolino** [comment]
I created a PR https://github.com/PrestaShop/PrestaShop/pull/26824

**Comment by matks** [comment]
> I wonder why something deprecated is used by `validateOrder` function in `classes/PaymentModule.php`. Any module that uses such functionality, as the MercadoPago official partner plugin, is going to fail under this configuration.

Hi @mpaolino  it's very simple: we're buried with work to do. Hundred of things to do, so it's hard to keep up. That's why we welcome any help such as your PR thanks 👍 

I reopen this issue but we can only validate your PR https://github.com/PrestaShop/PrestaShop/pull/26824 if we have a _clear, reproducible scenario_ we can follow to observe the bug. I can see the issue in the code, but in order to be able to explore it, to verify any side-effects, we need a complete understanding of it, and so does the QA team. That's why a list of steps "1. do this 2. do then that 3. we should see ... but instead we should see ..."  is a must-have needed for us to verify the bug 😉  .

**Comment by mpaolino** [comment]
Steps to reproduce the bug:

1. - Install clean Prestashop
1.1 - Configure Prestashop tu use a single currency with decimals value "0".

2. - Install [MercadoPago 4.8.0 module ](https://addons.prestashop.com/es/pago-tarjeta-carteras-digitales/23962-mercado-pago.html)
2.1 - Configure MercadoPago plugin with:
2.1.1 - Country: Uruguay
2.1.2 - Public Key: TEST-f9ea27ea-8d3b-4ebe-b614-648e7c85656d
2.1.3 - Access token: TEST-3885957295412805-120315-3cc32a81f07c5cbfa0d122dbe4f04547-19099354
2.2 - Select Visa and Master as valid card payment options under "CHECKOUT PRO" tab.
2.3 - Under Advanced Configuration (Configuracion Avanzada in Spanish) select all options as per screenshot.

![Screenshot from 2021-12-06 16-32-16](https://user-images.githubusercontent.com/757973/144910546-205d7ced-3ac3-42d5-a36b-7a511689920e.png)
![Screenshot from 2021-12-06 16-32-08](https://user-images.githubusercontent.com/757973/144910552-0e165f84-5cba-4c69-996a-7d0599a7e717.png)
![Screenshot from 2021-12-06 16-31-41](https://user-images.githubusercontent.com/757973/144910558-7f1cffde-5ad8-47b4-a6f7-6d13ceb39167.png)

3. Create an publish product with price lower than $1000
4. Create and publish product with price higher than $1000
5. Purchase product with price lower than $1000. Should work ok.
6. Purchase product with price higher than $1000. **This will fail and set order to Payment Error state.**

Credit card test values (from https://www.mercadopago.com.ar/developers/es/guides/online-payments/checkout-pro/test-integration):

Name: APRO APRO
Credit card number: 5031 7557 3453 0604
CVV: 123
Expiration: 11/2025


**Obtained result:**

Purchase in step 6 is set to Payment Error state. 

**Expect result:**

Both purchases in 5 and 6 are put in a valid state after purchase.

**Comment by mastudillot** [comment]
Updated steps to reproduce in the issue detail.

**Comment by mpaolino** [comment]
> Updated steps to reproduce in the issue detail.

Its a little more complex than previously thought, check the PR.

**Comment by ocio87** [comment]
Thanks @mpaolino for associating the error with this problem created, I hope it helps you. If you require verification of the possible solution, let me know and I will try to test it.

As I mentioned in the error, it also happens with the bank transfer module.

Greetings,

**Comment by mpaolino** [comment]
@florine2623 please recheck the initial reported steps. You made an order with the price less than 1000, you need to test it with a purchase higher than 1000.

**Comment by mpaolino** [comment]
@florine2623 How did you end up in the state "Payment accepted" with the default Payments by check module? Makes no sense to me as the default config of a fresh install is this:

![Screenshot from 2021-12-06 20-56-16](https://user-images.githubusercontent.com/757973/144941430-3343e7ba-8492-464a-844d-9a719982ca69.png)

Order should have ended up in status "Awaiting check payment" if no error would have happened. 


Anyways, to reproduce this bug you have 2 options. 

1. You follow the original instructions of @mastudillot with a minor correction to trigger the bug, you need to configure "Payments by check" module with:

![Screenshot from 2021-12-06 20-59-00](https://user-images.githubusercontent.com/757973/144941628-89817938-6bc9-4421-9188-c4ad2dde3c4b.png)

This will make sure the function that has the problem in PrestaShop is called and the error should manifest.

2. The other option is using my instructions, which are lengthier but will give same result.

**Comment by matks** [comment]
Hello everyone and thank you for the hard work for finding a clear and reproducible list of steps we can use to see the wrong behavior 🙏 

It might look like a lot of energy for not a good result, but I can assure you this is very useful to make sure we fix the issue because QA team will use the scenario to then poke around, change some settings and verify everything is good not only for the situation described but also in related scenarii.

**Comment by ocio87** [comment]
Hi,

Two days ago I decided to apply the change with what was reported in Pull requests https://github.com/PrestaShop/PrestaShop/pull/26824, I must say that the site is working correctly for me in Colombia, both with the bank transfer module and with the Wompi module, which is an external module. Before the proposed change it always throws me a payment error, since with the change the site works correctly with version 1.7.8.2.  It was an update from 1.7.7.8 to 1.7.8.2.

I appreciate the help, this error had stopped me installing from 1.7.8.0.

I clarify that it works correctly for me on the site I work, it does not mean that it will work with other sites, the complete approval of the team is important.

**Comment by mastudillot** [comment]
Hi @florine2623,

was it possible to verify the error? Do you need more detail on how to reproduce the problem?

**Comment by mpaolino** [comment]
Since it seems QA does not test with third party plugins (regardless if they trigger a core bug), here is a list of new steps to reproduce the issue. As in all the other cases, **you need to follow them to the letter if you expect to reproduce the bug**.

   1) Install a *clean* Prestashop version >= 1.7.8.0.
   2) Make sure you have installed latest "Payments by check"  module
   3) Create currency without decimal and set as default.
   4) Configure the order status "Awaiting check payment" status provided by "Payments by check" module with the option "Consider the order as validated" to true (enabled).
   5) Create an publish product with price lower than $1000.
   6) Create and publish product with price higher than $1000.
   7) Buy a product with the check payment method with a price less than $1000. It should work fine.
   8) Purchase product with check payment with price higher than $1000. This will fail and set order to Payment Error state.

**Comment by Guisardo** [comment]
I can confirm that the issue is still present at the version 1.7.8.2.
The constant _PS_PRICE_COMPUTE_PRECISION_ was deprecated:
https://github.com/PrestaShop/PrestaShop/blob/develop/config/config.inc.php#L138

But still is been used at the PaymentModule class:
https://github.com/PrestaShop/PrestaShop/blob/develop/classes/PaymentModule.php#L333
https://github.com/PrestaShop/PrestaShop/blob/develop/classes/PaymentModule.php#L1160

The quick workaround is to set the _PS_PRICE_COMPUTE_PRECISION_ constant in the config.inc.php file to the same precision as the one set on the store currency (this hack will work only if all the currencies have the same precision)

**Comment by josedusol** [comment]
Hello,

I'm experiencing this issue in a project with PS v1.7.8.1 and Cash on Delivery (COD) v1.0.6. In this case, the payment error occurs only with the currency UYU, and not with default currency USD. Both currencies are configured with 0 decimal.

Implementing pull request [#26824](https://github.com/PrestaShop/PrestaShop/pull/26824) solved the issue in this project.

Thanks @mpaolino.

**Comment by mpaolino** [comment]
> Hello,
> 
> I'm experiencing this issue in a project with PS v1.7.8.1 and Cash on Delivery (COD) v1.0.6. In this case, the payment error occurs only with the currency UYU, and not with default currency USD. Both currencies are configured with 0 decimal.
> 
> Implementing pull request [#26824](https://github.com/PrestaShop/PrestaShop/pull/26824) solved the issue in this project.
> 
> Thanks @mpaolino.

Arriba, me alegra que te haya resuelto.

**Comment by florine2623** [comment]
Hello all,

Sorry for the long answer.

I followed the lastest steps provided by @mpaolino :

> 1. Install a clean Prestashop version >= 1.7.8.0.
>2. Make sure you have installed latest "Payments by check" module
>3. Create currency without decimal and set as default.
>4. Configure the order status "Awaiting check payment" status provided by "Payments by check" module with the option "Consider the order as validated" to true (enabled).
>5. Create and publish product with price lower than $1000.
>6. Create and publish product with price higher than $1000.
>7. Buy a product with the check payment method with a price less than $1000. It should work fine.
>8. Purchase product with check payment with price higher than $1000. This will fail and set order to Payment Error state.

As mentioned in step 4 :
<img width="770" alt="Screenshot 2022-02-11 at 14 25 04" src="https://user-images.githubusercontent.com/16019289/153599538-32f2e5f5-b743-4d91-b8e0-7133c6ed9327.png">

I can reproduce your issue in PS 1.7.8.x and develop, as per my screen record :

https://user-images.githubusercontent.com/16019289/153600311-8791958d-93ed-4b12-874e-1a028c4c23f7.mov

It can't be reproduced in PS 1.7.7.x, so it is a regression :
<img width="1430" alt="Screenshot 2022-02-11 at 14 44 47" src="https://user-images.githubusercontent.com/16019289/153602316-d301e70e-5c6a-410f-ba27-46a258be8d97.png">


Thank you all for your comments and for the PR.
I will be working on it.

**Comment by MatShir** [comment]
PR has been merged for this issue, @PrestaShop/qa-team could you test it to be sure ? thanks :)

**Comment by AureRita** [comment]
Hi @MatShir ,

I tried to reproduce the Issue with prestashop v1.7.8.6 and the issue isn't here, as you can see : 

https://user-images.githubusercontent.com/100699445/175983501-ecfd7617-e2be-4358-b7f3-e0df45df3747.mp4

**Comment by ocio87** [comment]
This Bug was resolved in version 1.7.8.6 with the https://github.com/PrestaShop/PrestaShop/pull/26824
